### PR TITLE
Reduce allocations and transformations when creating a token.

### PR DESCRIPTION
**Pull request by brentschmaltz**
The purpose of this PR is to reduce the memory allocations when creating a JWS.
Strategy:
1. Use ArrayPools whenever possible.
2. Use a single MemoryStream at the highest level and make all writes into that stream. 
3. For .net 6+, add new API's to SignatureProviders to accept memory buffer to put signatures.
4. Make use of Encoding.GetBytes that take a memory stream.

With these changes benchmark shows a slightly faster creation with less than half the memory.

For 1000 token creations using SecurityTokenDescriptor 2.48MB vs. 7.2MB.

Additional work for composing tokens where the memory buffer is passed to CreateToken.

### CODE REVIEW

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs [hunk 1]
```diff
@@ -0,0 +1,1383 @@
...
+                            tokenDescriptor.TokenType);
+                    }
+                    else
+                    {
+                        return encodedChars.AsSpan(0, sizeOfEncodedHeader + sizeOfEncodedPayload + sizeOfEncodedSignature + 2).ToString();
+                    }
+                }
+                finally
+                {
+                    if (encodedChars is not null)
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs:275]: for these three, did you look at the average cases sizes and check if any of them make sense to stackalloc?

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs [hunk 2]
```diff
@@ -0,0 +1,1383 @@
...
+                Utf8JsonWriter writer = null;
+                char[] encodedChars = null;
+                byte[] asciiBytes = null;
+                byte[] signatureBytes = null;
+
+                try
+                {
+                    writer = new(utf8ByteMemoryStream, new JsonWriterOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping });
+
+                    WriteJwsHeader(
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs:200]: thoughts on a really small cache for encoded header values?

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs [hunk 3]
```diff
@@ -0,0 +1,1383 @@
...
+        internal static string CreateToken(
+            SecurityTokenDescriptor tokenDescriptor,
+            bool setdefaultTimesOnTokenCreation,
+            int tokenLifetimeInMinutes)
+        {
+            // The form of a JWS is: Base64UrlEncoding(UTF8(Header)) | . | Base64UrlEncoding(Payload) | . | Base64UrlEncoding(Signature)
+            // Where the Header is specifically the UTF8 bytes of the JSON, whereas the Payload encoding is not specified, but UTF8 is used by everyone.
+            // The signature is over ASCII(Utf8Bytes(Base64UrlEncoding(Header) | . | Base64UrlEncoding(Payload)))
+            // Since it is not known how large the JWS will be, a MemoryStream is used.
+            // An ArrayBufferWriter was benchmarked, while slightly faster, more memory is used and different code would be needed for 461+ and net6.0+
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs:185]: Probably for future work, but a IBufferWriter would fit better if we ever want to do a PooledByteBuffer

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs (hunk 2 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs:200]: @keegan-caruso you read my mind, if the key, algorithm are the same without any extras, we could have a cache.
I figured the PR was large enough, we can build all kinds of things on top of these changes.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs (hunk 3 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs:185]: @keegan-caruso yes, .net 8 will make that public.

File: benchmark/Microsoft.IdentityModel.Benchmarks/CreateTokenTests.cs [hunk 4]
```diff
@@ -19,16 +20,32 @@ public class CreateTokenTests
...
+                Issuer = Default.Issuer,
+                IssuedAt = now,
+                NotBefore = now,
                 SigningCredentials = KeyingMaterial.JsonWebKeyRsa256SigningCredentials,
             };
         }
 
         [Benchmark]
-        public string JsonWebTokenHandler_CreateToken() => _jsonWebTokenHandler.CreateToken(_tokenDescriptor);
+        public string JsonWebTokenHandler_CreateToken()
```
Comment (westin-m) [inline benchmark/Microsoft.IdentityModel.Benchmarks/CreateTokenTests.cs]: Should this instead be void?

File: benchmark/Microsoft.IdentityModel.Benchmarks/CreateTokenTests.cs [hunk 5]
```diff
@@ -19,16 +20,32 @@ public class CreateTokenTests
...
+        {
+            for (int i = 0; i< 1000; i++)
```
Comment (keegan-caruso) [inline benchmark/Microsoft.IdentityModel.Benchmarks/CreateTokenTests.cs]: this confuses me. Benchmark dotnet already does many iterations. Is this just to be a scalar on top of that? Why not just specify the iterations?

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs [hunk 6]
```diff
@@ -0,0 +1,1383 @@
...
+            if (tokenDescriptor.NotBefore.HasValue)
+            {
+                nbfSet = true;
+                writer.WritePropertyName(JwtPayloadUtf8Bytes.Nbf);
+                writer.WriteNumberValue(EpochTime.GetIntDate(tokenDescriptor.NotBefore.Value));
+            }
+
+            // Duplicates are resolved according to the following priority:
+            // SecurityTokenDescriptor.{Audience, Issuer, Expires, IssuedAt, NotBefore}, SecurityTokenDescriptor.Claims, SecurityTokenDescriptor.Subject.Claims
+            // SecurityTokenDescriptor.Claims are KeyValuePairs<string,object>, whereas SeSecurityTokenDescriptor.Subject.Claims are System.Security.Claims.Claim and are processed differently.
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs]: nit: "SeSecurityTokenDescriptor"

File: src/Microsoft.IdentityModel.Tokens/SupportedAlgorithms.cs [hunk 7]
```diff
@@ -358,5 +358,46 @@ internal static bool IsSupportedSymmetricAlgorithm(string algorithm)
                 || SymmetricKeyWrapAlgorithms.Contains(algorithm)
                 || SymmetricSigningAlgorithms.Contains(algorithm);
         }
+
+        /// <summary>
+        /// Returns the maximum size in bytes for a supported signature algorithms.
+        /// The key size affects the signature size for asymmetric algorithms.
+        /// </summary>
+        /// <param name="algorithm"></param>
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/SupportedAlgorithms.cs]: nit: empty params

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs [hunk 8]
```diff
@@ -0,0 +1,1383 @@
...
+                    writer?.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// A <see cref="SecurityTokenDescriptor"/> can contain claims from multiple locations.
+        /// This method consolidates the claims and adds default times {exp, iat, nbf} if needed.
+        /// </summary>
+        /// <param name="writer"></param>
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs]: empty params

**Review by keegan-caruso — APPROVED** [review]
Just a few small comments

File: benchmark/Microsoft.IdentityModel.Benchmarks/CreateTokenTests.cs (hunk 5 above)
Comment (brentschmaltz) [reply benchmark/Microsoft.IdentityModel.Benchmarks/CreateTokenTests.cs]: @keegan-caruso good point, i was just checking will remove this code.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs (hunk 1 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.CreateToken.cs:275]: @keegan-caruso another great idea.
Let's put it in our TODO list.

**Pull request merged**